sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -1027,23 +1028,30 @@ _pi_queue::resetCommandList(pi_command_list_ptr_t CommandList,
   }
 
   auto &EventList = CommandList->second.EventList;
-  // Check if standard commandlist
-  if (CommandList->second.ZeFence != nullptr) {
+  // Check if standard commandlist or fully synced in-order queue.
+  // If one of those conditions is met then we are sure that all events are
+  // completed so we don't need to check event status.
+  if (!CheckStatus || CommandList->second.ZeFence != nullptr ||
+      (isInOrderQueue() && !LastCommandEvent)) {
     // Remember all the events in this command list which needs to be
     // released/cleaned up and clear event list associated with command list.
     std::move(std::begin(EventList), std::end(EventList),
               std::back_inserter(EventListToCleanup));
     EventList.clear();
-  } else {
+  } else if (!isDiscardEvents()) {
     // For immediate commandlist reset only those events that have signalled.
+    // If events in the queue are discarded then we can't check their status.
     for (auto it = EventList.begin(); it != EventList.end();) {
       std::scoped_lock<pi_shared_mutex> EventLock((*it)->Mutex);
       ze_result_t ZeResult =
-          ZE_CALL_NOCHECK(zeEventQueryStatus, ((*it)->ZeEvent));
+          (*it)->Completed ? ZE_RESULT_SUCCESS : ZE_CALL_NOCHECK(zeEventQueryStatus, ((*it)->ZeEvent));
       if (ZeResult == ZE_RESULT_SUCCESS) {
         EventListToCleanup.push_back(std::move((*it)));
         it = EventList.erase(it);
       } else {
+        // Break early as soon as we found first incomplete event because next
+        // events are submitted even later.
+        break;
```

Comment:
It is a bit weird to have the it++ after the break statement because it will never be executed.
More importantly, only for an in-order queue is it true that finding an unset event in the list implies
all later events will also be unset. For out-of-order queues, events may have been added to the list in
increasing time sequence but they can complete in any order.

Reply:
Removed redundant it++, thanks. So, idea is that events which are submitted earlier are going to finish earlier. I.e. here we assume that if we found some event which is not yet finished then all next events are unlikely to be finished.
Just to avoid checking status of the events which are unlikely to be finished. But it is guaranteed that we will cleanup all events eventually during queue sync and release (where we do sync as well).